fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -238,4 +249,88 @@ private static void setDeprecated(DescriptorProtos.FieldDescriptorProto.Builder
+        if (MetaDataProtoEditor.hasUnion(fileDescriptor)) {
+            return fileDescriptor;
+        }
+        DescriptorProtos.FileDescriptorProto fileDescriptorProto = fileDescriptor.toProto();
+        DescriptorProtos.FileDescriptorProto.Builder fileBuilder = fileDescriptorProto.toBuilder();
+        fileBuilder.addMessageType(MetaDataProtoEditor.createDefaultUnion(fileDescriptor));
+        try {
+            return Descriptors.FileDescriptor.buildFrom(fileBuilder.build(), fileDescriptor.getDependencies().toArray(new Descriptors.FileDescriptor[0]));
+        } catch (Descriptors.DescriptorValidationException e) {
+            throw new MetaDataException("Failed to add a default union", e);
+        }
+    }
+
+    @Nonnull
+    private static DescriptorProtos.DescriptorProto.Builder createDefaultUnion(@Nonnull Descriptors.FileDescriptor fileDescriptor) {
+        DescriptorProtos.DescriptorProto.Builder unionMessageType = DescriptorProtos.DescriptorProto.newBuilder();
+        unionMessageType.setName(RecordMetaDataBuilder.DEFAULT_UNION_NAME);
+        for (Descriptors.Descriptor messageType : fileDescriptor.getMessageTypes()) {
+            RecordMetaDataOptionsProto.RecordTypeOptions.Usage messageTypeUsage = getMessageTypeUsage(messageType.toProto());
+            if (messageTypeUsage != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
+                addFieldToUnion(unionMessageType, messageType.getName(), messageType.getFullName());
+            }
+        }
+        return unionMessageType;
+    }
+
+    /**
+     * Creates a default union descriptor for the given file descriptor and a base union descriptor. It adds all of the
+     * non-{@code NESTED} message types that exist in the base union to the synthetic union.
+     * @param fileDescriptor the file descriptor to create a union for
+     * @param baseUnionDescriptor the base union descriptor
+     * @return the builder for the union
+     */
+    @Nonnull
+    @API(API.Status.INTERNAL)
+    public static DescriptorProtos.DescriptorProto.Builder createSyntheticUnion(@Nonnull Descriptors.FileDescriptor fileDescriptor, @Nonnull Descriptors.Descriptor baseUnionDescriptor) {
+        DescriptorProtos.DescriptorProto.Builder unionMessageType = DescriptorProtos.DescriptorProto.newBuilder();
+        unionMessageType.setName(RecordMetaDataBuilder.DEFAULT_UNION_NAME);
+        for (Descriptors.FieldDescriptor field : baseUnionDescriptor.getFields()) {
+            Descriptors.Descriptor messageType = fileDescriptor.findMessageTypeByName(field.getMessageType().getName());
+            RecordMetaDataOptionsProto.RecordTypeOptions.Usage messageTypeUsage = getMessageTypeUsage(messageType.toProto());
+            if (messageTypeUsage != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
+                DescriptorProtos.FieldDescriptorProto.Builder fieldBuilder = DescriptorProtos.FieldDescriptorProto.newBuilder()
```

Comment:
Why isn't this initialized with `field.toProto().toBuilder()`? I think the `type_name` still needs to be changed, but I think that would naturally preserve the number, the options, etc. (and anything else we forgot to carry over). (Well, this would actually need to clear out the `oneof_index`, I think, unless we carry over the `OneOf` descriptions from the base union descriptor to the new union message type.)

Reply:
I misunderstood your original comment. We are disallowing `oneof`, do we still need to this?

Reply:
Where are we disallowing `oneof`?

Reply:
Ah, it is in `addFieldToUnion`, but that's no longer being called in this method, I don't think.

Reply:
In `addFieldToUnion` for the stored meta-data. Since the stored union doesn't have it, we won't have it here too.

Reply:
I see. And then I see you added a check in this method to make sure there are no one ofs there, too.

I think that's fine. In theory, at some point, we might want to put one ofs everywhere that we can, but that's not this day.